File: problems/2047-number-of-valid-words-in-a-sentence/solution.c
int countValidWords(char* sentence) {
    int count = 0;
    int i = 0;
    while (sentence[i] != '\0') {
        while (sentence[i] == ' ') {
            i++;
        }
        if (sentence[i] == '\0') {
            break;
        }
        int start = i;
        while (sentence[i] != ' ' && sentence[i] != '\0') {
            i++;
        }
        int end = i - 1;
        int valid = 1;
        int hyphenCount = 0;
        int punctuationCount = 0;

        for (int j = start; j <= end; j++) {
            if (isdigit(sentence[j])) {
                valid = 0;
                break;
            } else if (sentence[j] == '-') {
                hyphenCount++;
                if (hyphenCount > 1) {
                    valid = 0;
                    break;
                }
                if (j == start || j == end) {
                    valid = 0;
                    break;
                }
                if (!islower(sentence[j - 1]) || !islower(sentence[j + 1])) {
                    valid = 0;
                    break;
                }
            } else if (sentence[j] == '!' || sentence[j] == '.' || sentence[j] == ',') {
                punctuationCount++;
                if (punctuationCount > 1) {
                    valid = 0;
                    break;
                }
                if (j != end) {
                    valid = 0;
                    break;
                }
            } else if (!islower(sentence[j]) && sentence[j] != ' ' && sentence[j] != '\0') {
                valid = 0;
                break;
            }
        }
        if (valid) {
            count++;
        }
    }
    return count;
}

File: problems/2047-number-of-valid-words-in-a-sentence/solution.py
class Solution:
    def countValidWords(self, sentence: str) -> int:
        count = 0
        for token in sentence.split():
            if not token:
                continue
            valid = True
            hyphen_count = 0
            punctuation_count = 0
            n = len(token)
            for i in range(n):
                if token[i].isdigit():
                    valid = False
                    break
                elif token[i] == '-':
                    hyphen_count += 1
                    if hyphen_count > 1:
                        valid = False
                        break
                    if i == 0 or i == n - 1:
                        valid = False
                        break
                    if not token[i - 1].islower() or not token[i + 1].islower():
                        valid = False
                        break
                elif token[i] in ['!', '.', ',']:
                    punctuation_count += 1
                    if punctuation_count > 1:
                        valid = False
                        break
                    if i != n - 1:
                        valid = False
                        break
            if valid:
                count += 1
        return count